Date: Fri, 3 Mar 2023 16:32:36 +0800
Subject: crypto: rng - Use a different crypto_rng for reseeding

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=2174928

Upstream Status: RHEL only

Currently, passing GRND_RANDOM flag to getrandom(2) in FIPS mode will
cause reseeding of the crypto_default_rng which will then slow down
other callers of getrandom(2) without the GRND_RANDOM flag. To avoid that
slowdown, a separate crypto_reseed_rng (DRBG) is added for reseeding.

Now those callers with the GRND_RANDOM flag set will reseed the
crypto_reseed_rng and retrieve random data from it.

This also fixes a bug where a large-enough GRND_RANDOM read could
be broken into three or more reads, in which case a separate user
request without GRND_RANDOM may come through in the middle without
triggering any reseeds.

Move the two RNGs onto separate cachelines to avoid interference.

Signed-off-by: Herbert Xu <herbert.xu@redhat.com>
(cherry picked from RHEL commit 746aec2f)
[resolve minor conflict in crypto_devrandom_read - our distro did not
port some changes that are ultimately removed by this patch]

SIM: https://i.amazon.com/D179864600
cr: https://code.amazon.com/reviews/CR-176347405

diff --git a/crypto/rng.c b/crypto/rng.c
--- a/crypto/rng.c
+++ b/crypto/rng.c
@@ -26,7 +26,9 @@
 
 #include "internal.h"
 
-static DEFINE_MUTEX(crypto_default_rng_lock);
+static ____cacheline_aligned_in_smp DEFINE_MUTEX(crypto_reseed_rng_lock);
+static struct crypto_rng *crypto_reseed_rng;
+static ____cacheline_aligned_in_smp DEFINE_MUTEX(crypto_default_rng_lock);
 struct crypto_rng *crypto_default_rng;
 EXPORT_SYMBOL_GPL(crypto_default_rng);
 static int crypto_default_rng_refcnt;
@@ -109,12 +111,12 @@ struct crypto_rng *crypto_alloc_rng(const char *alg_name, u32 type, u32 mask)
 }
 EXPORT_SYMBOL_GPL(crypto_alloc_rng);
 
-static int __crypto_get_default_rng(void)
+static int crypto_get_rng(struct crypto_rng **rngp)
 {
 	struct crypto_rng *rng;
 	int err;
 
-	if (!crypto_default_rng) {
+	if (!*rngp) {
 		rng = crypto_alloc_rng("stdrng", 0, 0);
 		err = PTR_ERR(rng);
 		if (IS_ERR(rng))
@@ -126,11 +128,9 @@ static int __crypto_get_default_rng(void)
 			return err;
 		}
 
-		crypto_default_rng = rng;
+		*rngp = rng;
 	}
 
-	crypto_default_rng_refcnt++;
-
 	return 0;
 }
 
@@ -139,7 +139,9 @@ int crypto_get_default_rng(void)
 	int err;
 
 	mutex_lock(&crypto_default_rng_lock);
-	err = __crypto_get_default_rng();
+	err = crypto_get_rng(&crypto_default_rng);
+	if (!err)
+		crypto_default_rng_refcnt++;
 	mutex_unlock(&crypto_default_rng_lock);
 
 	return err;
@@ -155,16 +157,17 @@ void crypto_put_default_rng(void)
 EXPORT_SYMBOL_GPL(crypto_put_default_rng);
 
 #if defined(CONFIG_CRYPTO_RNG) || defined(CONFIG_CRYPTO_RNG_MODULE)
-int crypto_del_default_rng(void)
+static int crypto_del_rng(struct crypto_rng **rngp, int *refcntp,
+		      struct mutex *lock)
 {
 	int err = -EBUSY;
 
-	mutex_lock(&crypto_default_rng_lock);
-	if (crypto_default_rng_refcnt)
+	mutex_lock(lock);
+	if (refcntp && *refcntp)
 		goto out;
 
-	crypto_free_rng(crypto_default_rng);
-	crypto_default_rng = NULL;
+	crypto_free_rng(*rngp);
+	*rngp = NULL;
 
 	err = 0;
 
@@ -173,6 +176,14 @@ int crypto_del_default_rng(void)
 
 	return err;
 }
+
+int crypto_del_default_rng(void)
+{
+	return crypto_del_rng(&crypto_default_rng, &crypto_default_rng_refcnt,
+			      &crypto_default_rng_lock) ?:
+	       crypto_del_rng(&crypto_reseed_rng, NULL,
+			      &crypto_reseed_rng_lock);
+}
 EXPORT_SYMBOL_GPL(crypto_del_default_rng);
 #endif
 
@@ -229,6 +240,7 @@ EXPORT_SYMBOL_GPL(crypto_unregister_rngs);
 static ssize_t crypto_devrandom_read(void __user *buf, size_t buflen,
 				     bool reseed)
 {
+	struct crypto_rng *rng;
 	u8 tmp[256];
 	ssize_t ret;
 
@@ -236,63 +248,38 @@ static ssize_t crypto_devrandom_read(void __user *buf, size_t buflen,
 		return 0;
 
 	if (reseed) {
-		int err;
-		int i;
 		u32 flags = 0;
 
 		/* If reseeding is requested, acquire a lock on
-		 * crypto_default_rng so it is not swapped out until
+		 * crypto_reseed_rng so it is not swapped out until
 		 * the initial random bytes are generated.
 		 *
 		 * The algorithm implementation is also protected with
 		 * a separate mutex (drbg->drbg_mutex) around the
 		 * reseed-and-generate operation.
 		 */
-		mutex_lock(&crypto_default_rng_lock);
+		mutex_lock(&crypto_reseed_rng_lock);
 
 		/* If crypto_default_rng is not set, it will be seeded
 		 * at creation in __crypto_get_default_rng and thus no
 		 * reseeding is needed.
 		 */
-		if (crypto_default_rng)
+		if (crypto_reseed_rng)
 			flags |= CRYPTO_TFM_REQ_NEED_RESEED;
 
-		ret = __crypto_get_default_rng();
+		ret = crypto_get_rng(&crypto_reseed_rng);
 		if (ret) {
-			mutex_unlock(&crypto_default_rng_lock);
+			mutex_unlock(&crypto_reseed_rng_lock);
 			return ret;
 		}
 
-		crypto_tfm_set_flags(crypto_rng_tfm(crypto_default_rng),
-				     flags);
-
-		i = min_t(int, buflen, sizeof(tmp));
-		err = crypto_rng_get_bytes(crypto_default_rng, tmp, i);
-
-		crypto_tfm_clear_flags(crypto_rng_tfm(crypto_default_rng),
-				       flags);
-
-		mutex_unlock(&crypto_default_rng_lock);
-		if (err) {
-			ret = err;
-			goto out;
-		}
-
-		if (copy_to_user(buf, tmp, i)) {
-			ret = -EFAULT;
-			goto out;
-		}
-
-		buflen -= i;
-		buf += i;
-		ret += i;
-
-		if (!buflen)
-			goto out;
+		rng = crypto_reseed_rng;
+		crypto_tfm_set_flags(crypto_rng_tfm(rng), flags);
 	} else {
 		ret = crypto_get_default_rng();
 		if (ret)
 			return ret;
+		rng = crypto_default_rng;
 	}
 
 	for (;;) {
@@ -300,7 +287,7 @@ static ssize_t crypto_devrandom_read(void __user *buf, size_t buflen,
 		int i;
 
 		i = min_t(int, buflen, sizeof(tmp));
-		err = crypto_rng_get_bytes(crypto_default_rng, tmp, i);
+		err = crypto_rng_get_bytes(rng, tmp, i);
 		if (err) {
 			ret = err;
 			break;
@@ -325,8 +312,10 @@ static ssize_t crypto_devrandom_read(void __user *buf, size_t buflen,
 		}
 	}
 
- out:
-	crypto_put_default_rng();
+	if (reseed)
+		mutex_unlock(&crypto_reseed_rng_lock);
+	else
+		crypto_put_default_rng();
 	memzero_explicit(tmp, sizeof(tmp));
 
 	return ret;
